File: backend/app/schemas/chatbot.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field


class Language(str, Enum):
    ENGLISH = "en"
    SWAHILI = "sw"


class MessageRole(str, Enum):
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"


class Message(BaseModel):
    role: MessageRole
    content: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    metadata: Optional[Dict[str, Any]] = None

    model_config = {
        "extra": "forbid"
    }


class ChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=2000)
    conversation_id: Optional[str] = None
    language: Language = Language.ENGLISH
    user_id: Optional[str] = None
    context: Optional[Dict[str, Any]] = None

    model_config = {
        "extra": "forbid"
    }


class ChatResponse(BaseModel):
    conversation_id: str
    message: str
    language: Language
    confidence: float = Field(..., ge=0.0, le=1.0)
    intent: Optional[str] = None
    entities: Optional[List[Dict[str, Any]]] = None
    sentiment: Optional[str] = None
    response_time: float
    metadata: Optional[Dict[str, Any]] = None

    model_config = {
        "extra": "forbid"
    }


class ConversationCreate(BaseModel):
    initial_message: str = Field(..., min_length=1, max_length=2000)
    language: Language = Language.ENGLISH
    user_id: Optional[str] = None
    context: Optional[Dict[str, Any]] = None

    model_config = {
        "extra": "forbid"
    }


class ConversationResponse(BaseModel):
    conversation_id: str
    messages: List[Message]
    language: Language
    created_at: datetime
    updated_at: datetime
    user_id: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None

    model_config = {
        "extra": "forbid"
    }


class ConversationHistory(BaseModel):
    conversation_id: str
    messages: List[Message]
    total_messages: int
    created_at: datetime
    updated_at: datetime

    model_config = {
        "extra": "forbid"
    }


class ChatbotFeedback(BaseModel):
    conversation_id: str
    message_id: str
    rating: int = Field(..., ge=1, le=5)
    feedback: Optional[str] = None
    helpful: bool = True

    model_config = {
        "extra": "forbid"
    }


class ChatbotAnalytics(BaseModel):
    total_conversations: int
    total_messages: int
    average_response_time: float
    average_confidence: float
    language_distribution: Dict[str, int]
    top_intents: List[Dict[str, Any]]
    sentiment_distribution: Dict[str, int]

    model_config = {
        "extra": "forbid"
    }

File: kipesa-frontend/types/chatbot.ts
export enum Language {
  ENGLISH = "en",
  SWAHILI = "sw"
}

export enum MessageRole {
  USER = "user",
  ASSISTANT = "assistant",
  SYSTEM = "system"
}

export interface Message {
  role: MessageRole
  content: string
  timestamp: string
  metadata?: Record<string, any>
}

export interface ChatRequest {
  message: string
  conversation_id?: string
  language: Language
  user_id?: string
  context?: Record<string, any>
}

export interface ChatResponse {
  conversation_id: string
  message: string
  language: Language
  confidence: number
  intent?: string
  entities?: Array<Record<string, any>>
  sentiment?: string
  response_time: number
  metadata?: Record<string, any>
}

export interface ConversationCreate {
  initial_message: string
  language: Language
  user_id?: string
  context?: Record<string, any>
}

export interface ConversationResponse {
  conversation_id: string
  messages: Message[]
  language: Language
  created_at: string
  updated_at: string
  user_id?: string
  metadata?: Record<string, any>
}

export interface ConversationHistory {
  conversation_id: string
  messages: Message[]
  total_messages: number
  created_at: string
  updated_at: string
}

export interface ChatbotFeedback {
  conversation_id: string
  message_id: string
  rating: number
  feedback?: string
  helpful: boolean
}

export interface ChatbotAnalytics {
  total_conversations: number
  total_messages: number
  average_response_time: number
  average_confidence: number
  language_distribution: Record<string, number>
  top_intents: Array<Record<string, any>>
  sentiment_distribution: Record<string, number>
}

export interface ChatState {
  conversation_id?: string
  messages: Message[]
  language: Language
  loading: boolean
  error: string | null
  typing: boolean
} 